File: exercism/ruby/sieve/sieve.rb
class Sieve
  def initialize(n)
    @limit = n
  end

  def primes
    return [] if @limit < 2

    range = 3..@limit
    [2] + range.step(2).select { |n| is_prime?(n) }
  end

  def is_prime?(n)
    i = 2
    while i <= Math.sqrt(n)
      return false if n % i == 0
      i += 1
    end

    true
  end
end

File: exercism/scala/sieve/src/main/scala/sieve.scala
object Sieve {

  def primesUpTo(limit: Int): Vector[Int] = limit match {
    case 0 | 1 => Vector.empty
    case _ =>
      def updateCache(n: Int): Set[Int] = (n to limit by n).filter(_ % n == 0).toSet

      def upTo(primes: Vector[Int], cache: Set[Int]): Vector[Int] = {
        val nextPrimeOpt = ((primes.last) to limit).toStream.filterNot(cache).headOption
        nextPrimeOpt.fold(primes)(p => upTo(primes :+ p, cache | updateCache(p)))
      }

      upTo(Vector(2), updateCache(2))
  }
}
